odoo/orm/models.py

```diff
@@ -3416,142 +3412,274 @@ def _check_access(self, operation: str) -> tuple[Self, Callable] | None:
+                field = self.env[ir_field.model]._fields[ir_field.name]
+                record = self.browse(json.loads(default.json_value))
+                raise UserError(_('Unable to delete %(record)s because it is used as the default value of %(field)s', record=record, field=field))
+
+        for field in many2one_fields:
+            model = self.env[field.model_name]
+            if field.ondelete == 'restrict':
+                if res := self.env.execute_query(SQL(
+                    """
+                    SELECT id, %(field)s
+                    FROM %(table)s
+                    WHERE %(field)s IS NOT NULL
+                    AND %(field)s @? %(jsonpath)s
+                    ORDER BY id
+                    LIMIT 1
+                    """,
+                    table=SQL.identifier(model._table),
+                    field=SQL.identifier(field.name),
+                    jsonpath=f"$.* ? ({' || '.join(f'@ == {id_}' for id_ in ids)})",
+                )):
+                    on_restrict_id, field_json = res[0]
+                    to_delete_id = next(iter(id_ for id_ in field_json.values()))
+                    on_restrict_record = model.browse(on_restrict_id)
+                    to_delete_record = self.browse(to_delete_id)
+                    raise UserError(_('You cannot delete %(to_delete_record)s, as it is used by %(on_restrict_record)s',
+                                        to_delete_record=to_delete_record, on_restrict_record=on_restrict_record))
+
+        def _post_hook():
+            # on delete set null/restrict for jsonb company dependent many2one
+            for field in many2one_fields:
+                model = self.env[field.model_name]
+                if field.ondelete == 'restrict':
+                    continue
+                res = self.env.execute_query(SQL(
+                    """
+                    UPDATE %(table)s
+                    SET %(field)s = (
+                        SELECT jsonb_object_agg(
+                            key,
+                            CASE
+                                WHEN value::int4 in %(ids)s THEN NULL
+                                ELSE value::int4
+                            END)
+                        FROM jsonb_each_text(%(field)s)
+                    )
+                    WHERE %(field)s IS NOT NULL AND %(field)s @? %(jsonpath)s
+                    RETURNING %(table)s."id"
+                    """,
+                    table=SQL.identifier(model._table),
+                    field=SQL.identifier(field.name, to_flush=field),
+                    ids=ids,
+                    jsonpath=f"$.* ? ({' || '.join(f'@ == {id_}' for id_ in ids)})",
+                ))
+                model.browse(id_ for id_, in res).invalidate_recordset([field.name])
 
-        self.check_access('unlink')
+        return _post_hook
+
+    @typing.final
+    def _delete_impacted_records(self, field: Field) -> BaseModel:
```

Comment:
Given the name, I thought it would call `unlink`.
```suggestion
    def _delete_collect_impacted_records(self, field: Field) -> BaseModel:
```